Orm/Xtensive.Orm.Tests/Issues/IssueGitHub0110_SimpleCommandProcessorOverridesOriginalException.cs

```diff
@@ -0,0 +1,166 @@
+{
+  [HierarchyRoot(InheritanceSchema.ConcreteTable)]
+  public class Bin : Entity
+  {
+    [Field, Key]
+    public long Id { get; private set; }
+
+    public Bin(Session session)
+      : base(session)
+    {
+    }
+  }
+
+  [HierarchyRoot(InheritanceSchema.ConcreteTable)]
+  [Index("Name", Unique = true)]
+  [Index(nameof(Active), nameof(N), nameof(M))]
+  public class InventoryBalance : Entity
+  {
+    [Field, Key]
+    public long Id { get; private set; }
+
+    [Field(Nullable = false)]
+    public Bin Bin { get; private set; }
+
+    [Field]
+    public bool Active { get; set; }
+
+    [Field]
+    public string Name { get; set; }
+
+    [Field]
+    public decimal N { get; set; }
+
+    [Field]
+    public decimal M { get; set; }
+
+    [Field(DefaultSqlExpression = "GETUTCDATE()")]
+    public DateTime ModifiedOn { get; set; }
+
+    public InventoryBalance(Session session, string name, Bin bin)
+      : base(session)
+    {
+      Name = name;
+      Active = true;
+      Bin = bin;
+    }
+  }
+
+  public class OperationState
+  {
+    public Exception CatchendException { get; set; }
+
+    public bool Ended { get; set; }
+
+    public Domain Domain { get; }
+
+    public OperationState(Domain domain)
+    {
+      Domain = domain;
+    }
+  }
+}
+
+namespace Xtensive.Orm.Tests.Issues
+{
+  public class IssueGitHub0110_SimpleCommandProcessorOverridesOriginalException : AutoBuildTest
+  {
+    protected override void CheckRequirements() => Require.ProviderIs(StorageProvider.SqlServer);
+
+    protected override DomainConfiguration BuildConfiguration()
+    {
+      var configuration = base.BuildConfiguration();
+      configuration.Types.Register(typeof(InventoryBalance).Assembly, typeof(InventoryBalance).Namespace);
+      configuration.UpgradeMode = DomainUpgradeMode.Recreate;
+
+      var defaultSessionConfig = configuration.Sessions.Default;
+      if (defaultSessionConfig == null) {
+        defaultSessionConfig = new SessionConfiguration(WellKnown.Sessions.Default, SessionOptions.ServerProfile);
+        configuration.Sessions.Add(defaultSessionConfig);
+      }
+
+      defaultSessionConfig.DefaultIsolationLevel = IsolationLevel.Snapshot;
+      defaultSessionConfig.DefaultCommandTimeout = 60 * 60; // 60 min.
+
+      defaultSessionConfig.BatchSize = 1; //force to use SimpleCommandProcessor
+
+      return configuration;
+    }
+
+    protected override void PopulateData()
+    {
+      using (var session = Domain.OpenSession())
+      using (var tx = session.OpenTransaction()) {
+        _ = new InventoryBalance(session, "aaa", new Bin(session));
+        tx.Complete();
+      }
+    }
+
+    [Test]
+    public void MainTest()
+    {
+      var task1State = new OperationState(Domain);
+      var task2State = new OperationState(Domain);
+
+      var task1 = new System.Threading.Tasks.Task(Outer, task1State);
+      var task2 = new System.Threading.Tasks.Task(Outer, task2State);
+      task1.Start();
+      task2.Start();
```

Comment:
I expect race conditions in this test sometimes. It is even possible task1 is completed before task2 started. 
I think it would be good to add some sort of synchronization here. Something like `ManualResetEvent` would help here.

Reply:
Thank you. This is good idea. I applied that.